tools/c7n_mailer/c7n_mailer/ldap_lookup.py

```diff
@@ -0,0 +1,146 @@
+# Copyright 2016 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import redis
+import re
+import shelve
+from ldap3 import Connection
+from ldap3.core.exceptions import LDAPSocketOpenError
+
+
+class LdapLookup(object):
+
+    def __init__(self, config, logger, get_connection=True):
+        # This is to allow us to instantiate this object and
+        # run the instance methods in unit tests
+        if get_connection:
+            self.connection = self.get_connection(
+                config.get('ldap_uri'),
+                config.get('ldap_bind_user', None),
+                config.get('ldap_bind_password', None)
+            )
+        self.log          = logger
+        self.base_dn      = config.get('ldap_bind_dn')
+        self.email_key    = config.get('ldap_email_key', 'mail')
+        manager_attr      = config.get('ldap_manager_attribute', 'manager')
+        self.attributes   = ['displayName', self.email_key, manager_attr]
+        self.uid_key      = config.get('ldap_username_attribute', 'sAMAccountName')
+        self.uid_regex    = config.get('ldap_uid_regex', None)
+        self.cache_engine = config.get('cache_engine', None)
+        if self.cache_engine == 'redis':
+            redis_host = config.get('redis_host')
+            redis_port = int(config.get('redis_port', 6379))
+            self.caching = redis.StrictRedis(host=redis_host, port=redis_port, db=0)
+        elif self.cache_engine == 'localshelf':
+            self.caching = LocalShelf('/tmp/ldap.cache')
+
+    def get_connection(self, ldap_uri, ldap_bind_user, ldap_bind_password):
+        # note, if ldap_bind_user and ldap_bind_password are None
+        # an anonymous bind will be attempted.
+        try:
+            conn = Connection(
+                ldap_uri, user=ldap_bind_user, password=ldap_bind_password,
+                auto_bind=True,
+                receive_timeout=30,
+                auto_referrals=False,
+            )
+        except LDAPSocketOpenError:
+            conn = None
+        return conn
+
+    def get_email_from_metadata(self, ldap_user_metadata):
+        return ldap_user_metadata.get(self.email_key, None)
+
+    def user_metadata_has_manager(self, ldap_user_metadata):
+        return ldap_user_metadata.get('manager', None)
+
+    def get_metadata_from_dn(self, user_dn):
```

Comment:
Since the 'manager' default AD schema has the manager as a DN, or direct path to the tree on AD, we need a separate instance method to do looks from a DN, versus doing lookups from a UID.

Since these are separate methods for doing lookups, we also store the keys different. There are key/value pairs for DNs, and different key / same values for UIDs